sys/include/can/can.h

```diff
@@ -95,15 +95,54 @@ struct can_frame {
     uint8_t data[CAN_MAX_DLEN] __attribute__((aligned(8)));
 };
 
+/**
+ * @brief Controller Area Network filter type
+ *
+ * The different CAN controllers provide different CAN filter
+ * types. These types are provided by the SAMD5x family (please check
+ * chapter 39.9.5 and 39.9.6 of the SAMD5x/SAME5x family Datasheet).
+ *
+ * The current CAN filter types list is from the SAMD5x/SAME5x family.
+ *
+ * Can be extended by other filter types if provided
+ */
+typedef enum {
+    CAN_FILTER_TYPE_RANGE = 0x00,   /**< Range filter from Filter 1 to Filter 2 (Filter 2 > Filter 1) */
+    CAN_FILTER_TYPE_DUAL,           /**< Dual ID Filter (Filter 2 or Filter 1) */
+    CAN_FILTER_TYPE_CLASSIC,        /**< Classic Filter: Filter ID and Mask */
+    CAN_FILTER_TYPE_EXT_RANGE       /**< For extended filters, Range filter from Filter 1 to Filter 2 (Filter 2 > Filter 1) */
+} can_filter_type_t;
+
+/**
+ * @brief Controller Area Network filter configuration
+ *
+ * The different CAN controllers provide different CAN filter
+ * configurations. These types are provided by the SAMD5x family (please check
+ * chapter 39.9.5 and 39.9.6 of the SAMD5x/SAME5x family Datasheet).
+ *
+ * The current CAN filter configuration list is from the SAMD5x/SAME5x family.
+ *
+ * Can be extended by other filter configurations if provided
+ */
+typedef enum {
+    CAN_FILTER_DISABLE = 0x00,      /**< Disable Filter element */
+    CAN_FILTER_RX_0,                /**< Store message in reception mailbox Rx_0 if filter matches */
+    CAN_FILTER_RX_1,                /**< Store message in reception mailbox Rx_1 if filter matches */
+    CAN_FILTER_RX_REJECT,           /**< Reject message if filter matches */
+    CAN_FILTER_RX_PRIO,             /**< Set priority if filter matches */
+    CAN_FILTER_RX_PRIO_0,           /**< Set priority and store message in reception mailbox Rx_0 if filter matches */
+    CAN_FILTER_RX_PRIO_1,           /**< Set priority and store message in reception mailbox Rx_1 if filter matches */
+    CAN_FILTER_RX_STRXBUF           /**< Store message in the RX buffer or as debug message */
+} can_filter_conf_t;
+
 /**
  * @brief Controller Area Network filter
  */
 struct can_filter {
     canid_t can_id;    /**< CAN ID */
     canid_t can_mask;  /**< Mask */
-#if defined(MODULE_MCP2515)
-    uint8_t target_mailbox; /**< The mailbox to apply the filter to */
-#endif
+    can_filter_conf_t can_filter_conf;      /**< CAN filter configuration */
+    can_filter_type_t can_filter_type;      /**< CAN filter type */
 };
```

Comment:
Why not have a struct per CAN filter type? That would make use much more ergonomic:

```C
typedef struct {
    can_filter_type_t type;
    uint8_t mailbox;
    uint16_t flags;
} can_filter_common_t;
```

Then you can have 

```C
typedef struct {
    can_filter_common_t super;
    canid_t id;
    canid_t mask;
} can_filter_classic_t;

typedef struct {
    can_filter_common_t super;
    canid_t start;
    canid_t end;
} can_filter_range_t;

typedef struct {
    can_filter_common_t super;
    canid_t id[2];
} can_filter_dual_t;

…
```

Reply:
The CAN driver interface provides function mapping, which for setting filters is `int (*set_filter)(candev_t *dev, const struct can_filter *filter)`. Should I switch also that or use the the common CAN struct as a starting point to define the other structs in the driver itself?

Reply:
Yes you are free to change the internal driver API.
If you want to keep compatibility with existing applications, you could instead introduce a new `.set_filter_ext()` function.

Reply:
Okay would this be an option?
`typedef struct {
       filter_type;
       filter_conf;     /* This will define where the data will be redirected (either mailbox 0 or 1) */ 
} can_filter_t;`

and from this base structure (which is still compatible with the current driver APIs) I can get the child structure depending from the type as you mentioned above

Reply:
Of course also with definition of child structs added here to my suggestion

Reply:
I think I don't quite understand yet what you are proposing, but just try it out and push what you have in mind.

Reply:
Ah no sorry nevermind, my suggestion will still need the can IDs as struct members. But also with your suggestion, shouldn't we add the CAN_IDs to the `can_filter_common_t` struct so that the user can input them?

Reply:
if you have `can_filter_classic_t`  be `struct can_filter` and have a zero super ( which schould not be super by the way) map to a classic can filter this will work:
```
struct can_filter myfil = { .id= 0x123 , .mask = 0x456 };
candev->setf_filter(candev, &myfil);
```
the super (which should not be called super will be set to zeros)

and thus the most used filter would work instantly

Reply:
If I understood well the `can_filter_common` idea we will need always to get the CAN IDs from the `can_filter` struct, which will make the `set_filter` API be `set_filter(candev, &can_filter, &can_filter_common)` and apply the filter accordingly (we still can use the child proposed structs inside the API). Do you think that's a good approach?
If yes, I would propose then instead of `can_filter_common`struct to have the normal `can_filter`struct and add a struct holding the configuration and type of the filter i.e. `struct filter_mode { can_filter_type; can_filter_conf;};` And then having `set_filter(candev, &can_filter, &filter_mode)`

Reply:
this is what  @benpicco suggested: 

```
typedef struct {
    can_filter_type_t type;
    uint8_t mailbox;
    uint16_t flags;
} can_filter_common_t;

Then you can have

typedef struct {
    can_filter_common_t super;
    canid_t id;
    canid_t mask;
} can_filter_classic_t;

```

it would work like

```
can_filter_classic_t myfil = { . id=0x123, .mask=0x456, super={.type=CLASSIC, ...}};
candev->set_filter(candev, &myfil);
```

Reply:
@benpicco suggested to have `can_filter_common` as input for the `set_filter` API and then from it extract the type of filter to use, and not directly `can_filter_classic`, that's the point making me question the whole approach (the `can_filter_common` does not contain the CAN IDs). If you look at it you'll see that the `can_filter_classic` is in the end similar to the `can_filter` structure I suggested including the type and conf (mailbox). Besides we shouldn't use different CAN filter structures definitions and use it as input argument because that will lead also to multiple `_set_filter()` definitions

Reply:
The idea is that there will never be a 'naked' `can_filter_common_t`, but that it is always part of a child struct - that you can get with `container_of()`.

You can always cast `can_filter_classic_t` to `can_filter_common_t`

Reply:
```
typedef struct {
    can_filter_type_t type;
    uint8_t mailbox;
    uint16_t flags;
} can_filter_typespec_t;

typedef struct can_filter{
    can_filter_typespec_t typespec;
    canid_t id;
    canid_t mask;
} can_filter_t;
```
^^^ how a bout this naming @benpicco

Reply:
set_filter can still take `set_filter(candev_t *dev, const struct can_filter *filter);`

we can cast from `can_filter_t` or `can_filter_classic_t` as well as from `can_filter_common_t` / `can_filter_typespec_t`

Reply:
we would just not use the cast hiding macro `container_of(`